hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/ImmutableSegment.java

```diff
@@ -43,4 +48,127 @@ public KeyValueScanner getKeyValueScanner() {
     return new CollectionBackedScanner(getCellSet(), getComparator());
   }
 
+  public boolean isFlat() {
+    return isFlat;
+  }
+
+  protected ImmutableSegment(Segment segment) {
+    super(segment);
+    isFlat = false;
+  }
+
+  // C-tor by flattening other (old) segment
+  protected ImmutableSegment(ImmutableSegment oldSegment, MemStoreCompactorIterator iterator,
+      MemStoreLAB memStoreLAB, int numOfCells, long constantCellSizeOverhead) {
+
+    super(null,oldSegment.getComparator(),memStoreLAB,
+        CompactingMemStore.DEEP_OVERHEAD_PER_PIPELINE_FLAT_ARRAY_ITEM, constantCellSizeOverhead);
+    CellSet cs = this.createArrayBasedCellSet(numOfCells, iterator, false); // build the CellSet
+    this.setCellSet(cs);      // update the CellSet of the new Segment
+    isFlat = true;
+  }
+
+  // C-tor by compaction to CellArrayMap or CellChunkMap
+  protected ImmutableSegment(
+      final Configuration conf, CellComparator comparator, MemStoreCompactorIterator iterator,
```

Comment:
if we use the iterator in the context of a segment then it should be Called InMemoryCompacteor and InMemoryCompactorIterator